man/unipagecount.1
.TH UNIPAGECOUNT 1 "2007 Dec 31"
.SH NAME
unipagecount \- Count the assigned code points in a GNU Unifont .hex file
.SH SYNOPSIS
.br
.B unipagecount
[-ppagenum] [-h|-l]
.SH DESCRIPTION
.B unipagecount
reads a GNU Unifont .hex file from STDIN and prints a 16 by 16 grid of the
number of defined code points in each 256 character block within a Unicode
plane to STDOUT.  Code points proceed from left to right, then top to bottom.
In all planes, code points U+*FFFE and U+*FFFF are not expected in the input
hex file; they are reserved and always counted as being present in a plane.
.SH OPTIONS
.TP 12
-Pplanenum
Select a Unicode plane, from 0 through 16, inclusive.  If not specified,
.B unipagecount
defaults to Plane 0 (the Basic Multilingual Plane).
.TP
-ppagenum
Just print information on one 256 code point "page" rather than
the entire Basic Multilingual Plane.  This prints a 16 by 16 table with
an asterisk in every code point that has an assigned glyph.
.TP
-h
Print an HTML table with color-coded cell background colors
instead of a plain text table.
.TP
-l
[The letter "l"]: Print hyperlinks to font bitmaps in the HTML table.
To create the bitmaps themselves, use the
.B unihex2bmp
program.  The bitmaps are assumed to be in the directory "bmp/".
.SH FILES
*.hex GNU Unifont font files
.SH SEE ALSO
.BR bdfimplode(1),
.BR hex2bdf(1),
.BR hex2sfd(1),
.BR hexbraille(1),
.BR hexdraw(1),
.BR hexkinya(1),
.BR hexmerge(1),
.BR johab2ucs2(1),
.BR unibdf2hex(1),
.BR unibmp2hex(1),
.BR unicoverage(1),
.BR unidup(1),
.BR unifont(5),
.BR unifont-viewer(1),
.BR unifont1per(1),
.BR unifontchojung(1),
.BR unifontksx(1),
.BR unifontpic(1),
.BR unigencircles(1),
.BR unigenwidth(1),
.BR unihex2bmp(1),
.BR unihex2png(1),
.BR unihexfill(1),
.BR unihexgen(1),
.BR unipng2hex(1)
.SH AUTHOR
.B unipagecount
was written by Paul Hardy.
.SH LICENSE
.B unipagecount
is Copyright \(co 2007, 2014 Paul Hardy.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH BUGS
No known real bugs exist, except that this software does not perform
extensive error checking on its input files.  If they're not in the
format of the original GNU Unifont .hex file, all bets are off.


NEWS
2017-06-20: Release 10.0.01

   Updated for Unicode 10.0 release.

   New scripts added (introduced in Unicode 10.0):
     - U+0860..U+086F     Syriac Supplement
     - U+011A00..U+011A4F Zanabazar Square
     - U+011A50..U+011AAF Soyombo
     - U+011D00..U+011D5F Masaram Gondi

   Added quadruple-width glyphs U+9FD8..U+9FE9, updating some utility
   programs and documentation to allow this change.  The Perl scripts
   unihex2png, unipng2hex, and unifont-viewer do not support this new
   16-by-32 glyph size.

   Corrected several glyphs and made numerous changes in others,
   described in detail in the ChangeLog file.

2016-12-22: Release 9.0.06

   Refined about 50 Pikto CSUR glyphs.

   Corrected two Osage glyphs and one mathematical operator as noted
   in ChangeLog.

2016-12-10: Release 9.0.05

   Modified Cherokee glyphs to reflect changes in recent Cherokee font
   that the Unicode Consortium adopted for its code charts.  Removed
   serifs to make Cherokee glyphs easier to read.

   Modified Pikto CSUR glyphs after comparison with high-resolution scans
   of Basic Pikto book.

2016-10-29: Release 9.0.04

   Extended "{" and "}" by one pixel so they now come to a point in the center.
   Swapped U+2A05 and U+2A06 so they are now in the correct order.

2016-10-21: Release 9.0.03

   Added CSUR glyphs:
     - Aiha: added doubling combining glyph, omitted from the ConScript
       Unicode Registry (CSUR) code chart.
     - Tonal System: base 16 counting system devised by John William Nystrom
       in the mid-19th century.
     - Pikto: added the Pikto glyphs to Plane 15 Private Use Area, in the
       range specified in the CSUR.

   The giant picture of all of Plane 0 now includes the Plane 0 CSUR glyphs,
   although those are an optional part of Unifont in separate font files.

2016-08-27: Release 9.0.02

   Added 3 more CSUR scripts.  Modifications to some existing glyphs.

2016-07-02: Release 9.0.01

   Updates to support Unicode 9.0.0.

2015-06-28: Release 8.0.01

  Updates to support Unicode 8.0.0.

2014-10-23: Release 7.0.06

   A few new scripts were added:
      U+010350..U+01037F Old Permic
      U+01F650..U+01F67F Ornamental Dingbats
      U+01F780..U+01F7FF Geometrical Shapes Extended
      U+01F800..U+01F8FF Supplemental Arrows-C

   Various changes were made to Makefiles, Perl scripts, and one bash
   shell script to conform to Debian Policy Manual version 3.9.6.

   Several changes were made to Makefiles:
      * Use "set -e" at the start of multi-command shell invocations,
        to trap errors.
      * Use "&&" rather than ";" in multi-command shell invocations,
        to fail on errors.
      * install now uses "-s" flag for C program binaries.

   Other minor changes were made to scripts; see ChangeLog
   for details.

2014-10-17: Release 7.0.05

   Modified ASCII letter 'l' so it would have the same height as
   the other lower-case letters with ascenders; this is one pixel
   taller than the capital letters.  Lowered the horizontal stroke 
   in 't' by one pixel so it would line up with the horizontal
   stroke in 'f'; this is one pixel lower than the x-height of
   the ASCII letters.  Letters elsewhere in Plane 0 that were formed
   with 'k', 'l', or 't' were checked and modified where appropriate
   to conform with the new ASCII glyphs.

   Completed the Miscellaneous Symbols and Pictographs script in Plane 1,
   and modified Plane 1 Emoticons and Transport and Map Symbols scripts.

   Now create file font/compiled/plane01.html, showing a color-coded
   indication of completion of all of Unicode Plane 1 as an HTML
   table in a self-contained HTML file.  That and the compiled/index.html
   are the two HTML files that building the font creates.  The plane01.html
   file gives a color coding to show completion of each 256-glyph block,
   from 0% (pure red) to 100% (pure green).

2014-10-11: Release 7.0.04

   Added five more Supplemental Multilingual Plane (Plane 1) scripts
   and improved others thanks to several contributors.  There are now
   over 5,000 drawn assigned glyphs for Plane 1!

   Reorganized font/plane01 directory.  Several other changes; see
   ChangeLog file for details.

   unihex2sfd and unipagecount now handle all Unicode planes, not just
   Plane 0.


2014-07-01: Release 7.0.03

   Minor release for Debian.  Added a man page for unihexfill(1);
   Debian requires a man page for each program.

2014-06-22: Release 7.0.02

   Fixes a bug in font/Makefile: there were two "done" statements at the
   end of one "for" loop.  Also adds many new Supplemental Multilingual
   Plane (Plane 1) scripts.  See the ChangeLog file for details.

2014-06-21: Release 7.0.01

   Provides a glyph for every visible code point in the Unicode 7.0 Basic
   Mulitlingual Plane.  Also adds more scripts in the Unicode 7.0
   Supplemental Multilingual Plane.  See the ChangeLog file for full
   details.

   Note that the versioning scheme has changed.  Instead of a date, this
   package will use an incrementing minor version number in the future
   ("01", "02", etc.).

2014-02-14: Release 6.3.20140214

   This version adds the GNU Free Documentation License 1.3 to the "COPYING"
   file.

2014-02-04: Release 6.3.20140204

   This version added Qianqian Fang's name to the COPYRIGHT string now
   passed to Fontforge when the TTF fonts are built, for his contributions
   contained in the file "font/plane00/wqy.hex".

   This version also removes "<img>" links to icons on external websites from
   the "hangul/hangul-generation.html" web page, which were retrieved and
   stored locally when the original web page was saved in the "hangul/"
   directory.

2014-02-02: Release 6.3.20140202

   This version is the first designed to easily build multiple fonts.
   To accomplish this, font/Makefile and font/ttfsrc/Makefile were
   modified to pass FONT and COMBINING parameters, giving the names
   of the font file and its associated combining.txt file.


src/unipagecount.c
/*
   unipagecount - program to count the number of glyphs defined in each page
                  of 256 bytes, and print them in an 8 x 8 grid.  Input is
                  from stdin.  Output is to stdout.

   Synopsis: unipagecount < font_file.hex > count.txt
             unipagecount -phex_page_num < font_file.hex  -- just 256 points
             unipagecount -h < font_file.hex              -- HTML table
             unipagecount -l < font_file.hex              -- linked HTML table

   Author: Paul Hardy, unifoundry <at> unifoundry.com, December 2007
   
   Copyright (C) 2007, 2008, 2013 Paul Hardy

   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>

#define MAXBUF 256 /* maximum input line size */

int
main (int argc, char *argv[])
{

   char inbuf[MAXBUF]; /* Max 256 characters in an input line */
   int i, j;  /* loop variables */
   unsigned page;  /* unicode page (256 bytes wide) */
   unsigned unichar; /* unicode character */
   int pagecount[256] = {256 * 0};
   int onepage=0; /* set to one if printing character grid for one page */
   int pageno=0; /* page number selected if only examining one page */
   int html=0;   /* =0: print plain text; =1: print HTML */
   int links=0;  /* =1: print HTML links; =0: don't print links */
   void mkftable();  /* make (print) flipped HTML table */

   size_t strlen();

   if (argc > 1 && argv[1][0] == '-') {  /* Parse option */
      switch (argv[1][1]) {
         case 'p':  /* specified -p<hexpage> -- use given page number */
            sscanf (&argv[1][2], "%x", &pageno);
            if (pageno >= 0 && pageno <= 255) onepage = 1;
            break;
         case 'h':  /* print HTML table instead of text table */
            html = 1;
            break;
         case 'l':  /* print hyperlinks in HTML table */
            links = 1;
            html = 1;
            break;
      }
   }
   /*
      Initialize pagecount to account for noncharacters.
   */
   if (!onepage) {
      pagecount[0xfd] = 32;  /* for U+FDD0..U+FDEF */
      pagecount[0xff] = 2;   /* for U+FFFE, U+FFFF */
   }
   /*
      Read one line at a time from input.  The format is:

         <hexpos>:<hexbitmap>

      where <hexpos> is the hexadecimal Unicode character position
      in the range 00..FF and <hexbitmap> is the sequence of hexadecimal
      digits of the character, laid out in a grid from left to right,
      top to bottom.  The character is assumed to be 16 rows of variable
      width.
   */
   while (fgets (inbuf, MAXBUF-1, stdin) != NULL) {
      sscanf (inbuf, "%X", &unichar);
      page = unichar >> 8;
      if (onepage) { /* only increment counter if this is page we want */
         if (page == pageno) { /* character is in the page we want */
            pagecount[unichar & 0xff]++; /* mark character as covered */
         }
      }
      else { /* counting all characters in all pages */
         /* Don't add in noncharacters (U+FDD0..U+FDEF, U+FFFE, U+FFFF) */
         if (unichar < 0xfdd0 || (unichar > 0xfdef && unichar < 0xfffe))
            pagecount[page]++;
      }
   }
   if (html) {
      mkftable (pagecount, links);
   }
   else {  /* Otherwise, print plain text table */
      fprintf (stdout,
         "   0   1   2   3   4   5   6   7   8   9   A   B   C   D   E   F\n");
      for (i=0; i<0x10; i++) {
         fprintf (stdout,"%X ", i); /* row header */
         for (j=0; j<0x10; j++) {
            if (onepage) {
               if (pagecount[i*16+j])
                  fprintf (stdout," *  ");
               else
                  fprintf (stdout," .  ");
            }
            else {
               fprintf (stdout, "%3X ", pagecount[i*16+j]);
            }
         }
         fprintf (stdout,"\n");
      }
   
   }
   exit (0);
}


/*
   mkftable - function to create an HTML table to show flipped bmp files
              in a 16 by 16 grid.
*/

void
mkftable (int pagecount[256], int links)
{
   int i, j;
   int count;
   unsigned bgcolor;
   
   printf ("<html>\n");
   printf ("<body>\n");
   printf ("<table border=\"3\" align=\"center\">\n");
   printf ("  <tr><th colspan=\"16\" bgcolor=\"#ffcc80\">");
   printf ("GNU Unifont Page Coverage<br>(Green=100%%, Red=0%%)</th></tr>\n");
   for (i = 0x0; i <= 0xF; i++) {
      printf ("  <tr>\n");
      for (j = 0x0; j <= 0xF; j++) {
         count = pagecount[ (i << 4) | j ];
         
         /* print link in cell if links == 1 */
         if (i < 0xd || (i == 0xd && j < 0x8) || (i == 0xf && j > 0x8)) {
            /* background color is light green if completely done */
            if (count == 0x100) bgcolor = 0xccffcc;
            /* otherwise background is a shade of yellow to orange to red */
            else bgcolor = 0xff0000 | (count << 8) | (count >> 1);
            printf ("    <td bgcolor=\"#%06X\">", bgcolor);
            printf ("<a href=\"bmp/uni%X%X.bmp\">%X%X</a>", i, j, i, j);
            printf ("</td>\n");
         }
         else if (i == 0xd) {
            if (j == 0x8) {
               printf ("    <td align=\"center\" colspan=\"8\" bgcolor=\"#cccccc\">");
               printf ("<b>Surrogate Pairs</b>");
               printf ("</td>\n");
            }  /* otherwise don't print anything more columns in this row */
         }
         else if (i == 0xe) {
            if (j == 0x0) {
               printf ("    <td align=\"center\" colspan=\"16\" bgcolor=\"#cccccc\">");
               printf ("<b>Private Use</b>");
               printf ("</td>\n");
            }  /* otherwise don't print any more columns in this row */
         }
         else if (i == 0xf) {
            if (j == 0x0) {
               printf ("    <td align=\"center\" colspan=\"9\" bgcolor=\"#cccccc\">");
               printf ("<b>Private Use</b>");
               printf ("</td>\n");
            }
         }
      }
      printf ("  </tr>\n");
   }
   printf ("</table>\n");
   printf ("</body>\n");
   printf ("</html>\n");

   return;
}
